pages/en/zkapps/how-to-write-a-zkapp.mdx

```diff
@@ -207,121 +205,192 @@ For a full list, see the <DocLink copy="SnarkyJS reference" url="/zkapps/snarkyj
+In a method, you can use SnarkyJS commands like `Field.mul` and `Field.assertEquals` to define your custom logic.
+
+Later, we'll show how you can...
+
+- run a method off-chain
+- create a proof that it executed successfully
+- send that proof to the Mina network, to trigger actions like a state change or payment
+
+To get an idea what "successful execution" means, look at this line in our example above:
 
 ```ts
-super.deploy();
-this.balance.addInPlace(initialBalance);
+x.mul(2).assertEquals(5);
 ```
 
-You are free to add additional parameters, like the initial values of your state.
-State variables can be initialized during deployment using the `.set` method on `State`:
+Creating a proof for this method will only be possible if the input `x` satisfies the equation `x * 2 === 5`. This is what we call a "constraint".
+Magically, the proof can be checked without seeing `x` -- it's a _private input_.
+So, you can use `myMethod()` to prove that you understand modular arithmetic 😉 <!-- TODO: is this stupid? -->
+
+Jokes aside, the method above is not very meaningful yet. To make it more interesting, we need a way to interact with accounts, and record state on-chain.
+Check out the next section for more on that!
+
+One more note about private inputs: The method above has one input parameter, `x` of type `Field`. In general, arguments can be any of the built-in SnarkyJS type that you saw: `Bool`, `UInt64`, `PrivateKey`, etc. From now on, we will refer to those types as _circuit values_. <!-- TODO: we need a dedicated section about this, also introducing CircuitValue! -->
+
+<Alert kind="info">
+<!-- TODO -- too much? too early? I think little "deep dives" like this can be useful to answer questions that more advanced users often have after reading our docs,
+     and spread more understanding of Mina to the internet 
+-->
+
+Under the hood, every `@method` defines a zkSNARK circuit. From the cryptography standpoint, a smart contract is a collection of circuits, all of which are compiled into a single prover / verifier key. The proof says something to the effect of "I ran one of these methods, with some private input, and it produced this particular set of account updates". In SNARK terms, the account updates are the _public input_. The proof will only be accepted on the network if it verifies against the verifier key stored in the account. This ensures that indeed, the same code that the zkApp developer wrote also ran on the user's device -- thus, the account updates conform to the smart contract's rules.
+
+</Alert>
+
+##### On-chain state
+
+A smart contract can contain **on-chain state**, which is declared as a property
+on the class with the `@state` decorator:
 
 ```ts
-this.x.set(x);
+class HelloWorld extends SmartContract {
+  @state(Field) x = State<Field>();
+
+  // ...
+}
 ```
 
-Uninitialized values will be zero.
+Here, `x` is of type `Field`. Like with method inputs, only SnarkyJS circuit values can be used for state variables.
```

Comment:
Is there any other phrasing we could consider than `circuit values`? I know what you're saying and it's useful to mention, just thinking of Brandon mentioning that this is an implementation detail. If there is no easy alternative, this is ok.

Reply:
well, we have a class called `CircuitValue` which is pretty important to define you own custom circuit types. I was trying to be consistent with that. As long as we keep that I'd say we don't have to worry about avoiding the term

Reply:
Yeah, we've talked about renaming that when we can think of a better name. This is fine for now